packages/opentelemetry-scope-zone-peer-dep/src/ZoneScopeManager.ts

```diff
@@ -0,0 +1,393 @@
+        target.__ot_listeners[event] === undefined
+      ) {
+        return original.call(this, event, listener);
+      }
+      const events = target.__ot_listeners[event];
+      const patchedListener = events.get(listener);
+      events.delete(listener);
+      return original.call(this, event, patchedListener || listener);
+    };
+  }
+
+  /**
+   * Restores previous scope for certain zone
+   * @param zone
+   */
+  private _restorePreviousScope(zone: Zone) {
+    const zoneName = zone.name;
+    const scopes = this._scopes[zoneName];
+    if (scopes && scopes.length > 0) {
+      scopes.splice(scopes.length - 1, 1);
+    }
+  }
+
+  /**
+   * Returns the active scope
+   */
+  active(): unknown | undefined {
+    const activeZoneName = this._getActiveZoneName();
+
+    const active = this._activeScopeFromZoneName(activeZoneName);
+    if (active) {
+      return active;
+    }
+    if (this._enabled) {
+      return window;
+    }
+    return undefined;
+  }
+
+  /**
+   * Binds a the certain scope or the active one to the target function and then returns the target
+   * @param target
+   * @param scope A scope (span) to be bind to target
+   */
+  bind<T>(target: T | TargetWithEvents, scope?: unknown): T {
+    // if no specific scope to propagate is given, we use the current one
+    if (scope === undefined) {
+      scope = this.active();
+    }
+    if (typeof target === 'function') {
+      return this._bindFunction(target, scope);
+    } else if (isListenerObject(target)) {
+      this._bindListener(target, scope);
+    }
+    return (target as unknown) as T;
+  }
+
+  /**
+   * Disable the scope manager (clears all the scopes)
+   */
+  disable(): this {
+    const zoneNames = Object.keys(this._scopes);
+    for (const zoneName of zoneNames) {
+      this._cleanScopesFromZone(zoneName);
+    }
+    this._scopes = {};
+    this._enabled = false;
+    return this;
+  }
+
+  /**
+   * Enables the scope manager and creates a default(root) scope
+   */
+  enable(): this {
+    if (this._enabled) {
+      return this;
+    }
+    this._enabled = true;
+    return this;
+  }
+
+  /**
+   * Calls the callback function [fn] with the provided [scope].
+   *     If [scope] is undefined then it will use the active scope.
+   *     The scope will be set as active
+   * @param scope A scope (span) to be called with provided callback
+   * @param fn Callback function
+   */
+  with<T extends (...args: unknown[]) => ReturnType<T>>(
+    scope: unknown,
+    fn: () => ReturnType<T>
+  ): ReturnType<T> {
+    // if no scope use active from active zone
+    if (typeof scope === 'undefined' || scope === null) {
+      scope = this.active();
+    }
+
+    let zoneName: string;
+    let activeZone = this._getActiveZone();
+    if (!activeZone) {
+      zoneName = this._createZoneName();
+    } else {
+      zoneName = activeZone.name;
+    }
+
+    this._scopes[zoneName] = this._scopes[zoneName] || [];
+    this._scopes[zoneName].push(scope);
```

Comment:
Here what's happening is that if the user is already in a Zone that is associated with a scope, it pushes on that list - but that's not done in an async aware way.

So what if the user does something like this:
```
const rootSpan = webTracerWithZone.startSpan('rootSpan');
webTracerWithZone.withSpan(span1, () => {
  // A new Zone.js Zone is created because this is the first span.
  // Let's pretend the Zone is "Zone A".

  console.log('Current span is rootSpan', webTracerWithZone.getCurrentSpan() === rootSpan);

  const concurrentSpan1 = webTracerWithZone.startSpan('concurrentSpan1');
  const concurrentSpan2 = webTracerWithZone.startSpan('concurrentSpan2');

  webTracerWithZone.withSpan(concurrentSpan1, () => {
      // A concurrentSpan1 entry is pushed onto the list of scopes for Zone A.
      setTimeout(() => {
        // The concurrentSpan2 entry is the top entry for Zone A!!!! It's
        // Still waitin on its 200ms timer, so it will be the current span,
        // which is incorrect.
        console.log('Current span is incorrectly concurrentSpan2', webTracerWithZone.getCurrentSpan() === concurrentSpan2);
      }, 100);
  });

  webTracerWithZone.withSpan(concurrentSpan2, () => {
      // A new concurrentSpan2 entry is pushed onto the list of scopes for Zone A.
      setTimeout(() => {
        console.log('Current span is concurrentSpan2', webTracerWithZone.getCurrentSpan() === concurrentSpan2);
      }, 200);
  });
});
```

I would suggest that we should make each scope be its own Zone. I actually suspect that the same bug might exist in the async hooks scope manager, though I don't know async hooks as well.

I think making each scope its own Zone also makes it easier to get the current scope from a Zone. At its core it would be a line like this`Zone.current.get('_ot_Scope')` or similar. Then the `with` call would always involve creating a new Zone. You would need to do some testing around the hierarchical nature of Zones to make sure this does what you want.

Reply:
converted this into unit test